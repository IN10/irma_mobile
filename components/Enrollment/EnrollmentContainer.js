import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ReactTimeout from 'react-timeout';

import Enrollment from './Enrollment';
import { resetNavigation } from 'lib/navigation';
import { getLanguage } from 'lib/i18n';

const mapStateToProps = (state) => {
  const {
    enrollment: {
      error,
      status,
    }
  } = state;

  return {
    error,
    status,
  };
};

@ReactTimeout
@connect(mapStateToProps)
export default class EnrollmentContainer extends Component {

  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    error: PropTypes.object,
    navigation: PropTypes.object.isRequired,
    status: PropTypes.string.isRequired,
    setTimeout: PropTypes.func.isRequired,
  }

  static navigationOptions = Enrollment.navigationOptions;

  state = {
    email: null,
    validationForced: false,
    pin: null,
    disableRetry: false,
  }

  componentDidUpdate(prevProps) {
    const { status, navigation } = this.props;

    // When successfully enrolled, reset the route so we can't go back to EnrollmentTeaser
    // TODO: This creates an unwanted animation, but react-navigation doesn't seem to support
    // not displaying it, despite years of tickets. Only workaround seems to be react-navigation#1490
    // Consider first upgrading react-navigation before attempting this.
    if(prevProps.status !== status && status === 'success') {
      resetNavigation(navigation.dispatch, 'Enrollment');
    }
  }

  changeEmail(email) {
    this.setState({email});
  }

  changePin(pin) {
    this.setState({pin});
  }

  enroll({ pin, email }) {
    const { dispatch } = this.props;
    const language = getLanguage();

    // We take the passed (pin and) email value, because the user could've skipped while having
    // validly filled the email fields. So we record the final value here for retries
    this.setState({email, pin});

    dispatch({
      type: 'IrmaBridge.Enroll',
      email,
      pin,
      language,
    });
  }

  retryEnroll() {
    const { dispatch } = this.props;
    const { email, pin } = this.state;
    const language = getLanguage();

    // Disallow retry for three seconds
    this.setState({disableRetry: true});
    this.props.setTimeout(
      () => this.setState({disableRetry: false}),
      3000
    );

    dispatch({
      type: 'IrmaBridge.Enroll',
      email,
      pin,
      language
    });
  }

  navigateBack() {
    const { dispatch, navigation } = this.props;

    dispatch({
      type: 'Enrollment.Dismiss'
    });

    navigation.goBack();
  }

  navigateToDashboard() {
    const { navigation } = this.props;
    resetNavigation(navigation.dispatch, 'CredentialDashboard');
  }

  render() {
    const { error, status } = this.props;
    const { disableRetry, email, pin, validationForced } = this.state;

    return (
      <Enrollment
        disableRetry={disableRetry}
        changeEmail={::this.changeEmail}
        changePin={::this.changePin}
        email={email}
        enroll={::this.enroll}
        error={error}
        navigateBack={::this.navigateBack}
        navigateToDashboard={::this.navigateToDashboard}
        pin={pin}
        retryEnroll={::this.retryEnroll}
        status={status}
        validationForced={validationForced}
      />
    );
  }
}
