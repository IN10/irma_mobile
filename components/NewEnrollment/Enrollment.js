import React, { Component } from 'react';
import PropTypes from 'prop-types';

import {
  Button,
  Text,
  Container,
  Card,
  CardItem,
  View,
  Footer,
} from 'native-base';

import { namespacedTranslation } from 'lib/i18n';
import CollapsableForm from 'lib/form/CollapsableForm';
import PaddedContent from 'lib/PaddedContent';

export const t = namespacedTranslation('Enrollment');

export default class Enrollment extends Component {

  static propTypes = {
    changeEmail: PropTypes.func.isRequired,
    changePin: PropTypes.func.isRequired,
    email: PropTypes.string,
    enroll: PropTypes.func.isRequired,
    navigateToDashboard: PropTypes.func.isRequired,
    pin: PropTypes.string,
  }

  state = {
    emailFormCollapsed: true,
    pinFormCollapsed: false,
    showSuccess: false,
    validationForced: false,
  }

  renderIntro() {
    return (
      <Card>
        <CardItem>
          <Text>
            { t('.intro') }
          </Text>
        </CardItem>
      </Card>
    );
  }

  renderPinForm() {
    const { pin, changePin } = this.props;
    const { validationForced, pinFormCollapsed, showSuccess } = this.state;

    const next = () => {
      if(pin)
        this.setState({validationForced: false, pinFormCollapsed: true, emailFormCollapsed: false});
      else
        this.setState({validationForced: true});
    };

    return (
      <CollapsableForm
        testID="pinForm"
        headerText={ t('.step1.header') }

        onToggleCollapse={() => this.setState({pinFormCollapsed: !pinFormCollapsed})}
        collapsed={pinFormCollapsed}
        validationForced={validationForced}

        onNext={next}

        onChange={changePin}
        value={pin}
        locked={showSuccess}

        firstLabel={ t('.step1.label') }
        repeatLabel={ t('.step1.repeatLabel') }
        invalidMessage={ t('.step1.invalidMessage') }
        inputType="pin"
      >
        <Text>
          { t('.step1.text') }
        </Text>
      </CollapsableForm>
    );
  }

  renderEmailForm() {
    const { pin, email, changeEmail, enroll } = this.props;
    const { validationForced, emailFormCollapsed, showSuccess } = this.state;

    const next = () => {
      if(!email) {
        this.setState({validationForced: true});
        return;
      }

      this.setState({emailFormCollapsed: true, validationForced: false});

      if(pin) {
        enroll({ pin, email });
      } else
        this.setState({pinFormCollapsed: false});
    };

    const skip = () => {
      this.setState({emailFormCollapsed: true, validationForced: false});

      if(pin)
        enroll({ pin, email: null });
      else
        this.setState({pinFormCollapsed: false});
    };

    return (
      <CollapsableForm
        headerText={ t('.step2.header') }

        onToggleCollapse={() => this.setState({emailFormCollapsed: !emailFormCollapsed})}
        collapsed={emailFormCollapsed}
        validationForced={validationForced}

        onSkip={skip}
        onNext={next}

        onChange={changeEmail}
        value={email}
        locked={showSuccess}

        firstLabel={t('.step2.label')}
        repeatLabel={t('.step2.repeatLabel')}
        invalidMessage={ t('.step2.invalidMessage') }
        inputType="email"
      >
        <Text>
          { t('.step2.text') }
        </Text>
      </CollapsableForm>
    );
  }

  renderForm() {
    return (
      <View>
        { this.renderIntro() }
        { this.renderPinForm() }
        { this.renderEmailForm() }
      </View>
    );
  }

  renderSuccessCard() {
    return (
      <Card>
        <CardItem>
          <Text>
            { t('.success') }
          </Text>
        </CardItem>
      </Card>
    );
  }

  renderSuccessFooter() {
    const { navigateToDashboard } = this.props;

    return (
      <Footer style={{height: 60, paddingTop: 7}}>
        <Button primary onPress={navigateToDashboard}>
          <Text>{ t('.finish') }</Text>
        </Button>
      </Footer>
    );
  }

  render() {
    const { showSuccess } = this.state;

    return (
      <Container testID="Enrollment" style={{backgroundColor: '#E9E9EF'}}>
        <PaddedContent>
          { this.renderForm() }
          { !showSuccess ? null :
              this.renderSuccessCard()
          }
        </PaddedContent>
        { !showSuccess ? null :
            this.renderSuccessFooter()
        }
      </Container>
    );
  }
}
