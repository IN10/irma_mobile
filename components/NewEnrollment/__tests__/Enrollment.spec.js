/*global describe expect it*/
import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import Enrollment from '../';

const storeState = {
  enrollment: {status: 'started'}
};

describe('Enrollment', () => {

  it('renders as expected', () => {
    const wrapper = shallow(
      <Enrollment navigation={{}} />,
      { context: { store: configureStore()(storeState) } }
    );

    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

});
