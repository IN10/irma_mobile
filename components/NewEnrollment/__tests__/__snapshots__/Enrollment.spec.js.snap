// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enrollment renders as expected 1`] = `
<Styled(Container)
  style={
    Object {
      "backgroundColor": "#E9E9EF",
    }
  }
  testID="Enrollment"
>
  <KeyboardAwareComponent>
    <Styled(ViewNB)>
      <Styled(Card)>
        <Styled(CardItem)>
          <Styled(Text)>
            Welcome to IRMA! Please open your IRMA account by completing the two steps below.
          </Styled(Text)>
        </Styled(CardItem)>
      </Styled(Card)>
      <CollapsableForm
        collapsed={false}
        firstLabel="PIN"
        headerText="Step 1: Choose an IRMA PIN pin"
        inputType="pin"
        invalidMessage="Please enter a PIN of at least 5 digits"
        locked={false}
        onChange={[Function]}
        onNext={[Function]}
        onToggleCollapse={[Function]}
        repeatLabel="Repeat PIN"
        testID="pinForm"
        validationForced={false}
        value={null}
      >
        <Styled(Text)>
          Enter a PIN of at least 5 digits. You will need to enter your PIN every time you use IRMA. If you forget it, your attributes become unusable and you will have to open a new account.
        </Styled(Text)>
      </CollapsableForm>
      <CollapsableForm
        collapsed={true}
        firstLabel="E-mail address"
        headerText="Step 2: Add your email address"
        inputType="email"
        invalidMessage="Please enter a valid e-mail address"
        locked={false}
        onChange={[Function]}
        onNext={[Function]}
        onSkip={[Function]}
        onToggleCollapse={[Function]}
        repeatLabel="Repeat e-mail address"
        validationForced={false}
        value={null}
      >
        <Styled(Text)>
          You must associate an email address with your account. It allows you to login to your MyIRMA environment in case your phone is lost or stolen, so you can close your account.
        </Styled(Text)>
      </CollapsableForm>
    </Styled(ViewNB)>
  </KeyboardAwareComponent>
</Styled(Container)>
`;
